pkg/core/native_nep5.go

```diff
@@ -0,0 +1,234 @@
+}
+
+func (c *nep5TokenNative) Transfer(ic *interopContext, args []vm.StackItem) vm.StackItem {
+	from := toUint160(args[0])
+	to := toUint160(args[1])
+	amount := toInt64(args[2])
+	err := c.transfer(ic, from, to, amount)
+	return vm.NewBoolItem(err == nil)
+}
+
+func addrToStackItem(u *util.Uint160) vm.StackItem {
+	if u == nil {
+		return nil
+	}
+	return vm.NewByteArrayItem(u.BytesBE())
+}
+
+func (c *nep5TokenNative) emitTransfer(ic *interopContext, from, to *util.Uint160, amount int64) {
+	ne := state.NotificationEvent{
+		ScriptHash: c.hash,
+		Item: vm.NewArrayItem([]vm.StackItem{
+			vm.NewByteArrayItem([]byte("Transfer")),
+			addrToStackItem(from),
+			addrToStackItem(to),
+			vm.NewBigIntegerItem(amount),
+		}),
+	}
+	ic.notifications = append(ic.notifications, ne)
+}
+
+func (c *nep5TokenNative) transfer(ic *interopContext, from, to util.Uint160, amount int64) error {
+	if amount < 0 {
+		return errors.New("negative amount")
+	}
+
+	accFrom, err := ic.dao.GetAccountStateOrNew(from)
+	if err != nil {
+		return err
+	}
+
+	isEmpty := from.Equals(to) || amount == 0
+	inc := amount
+	if isEmpty {
+		inc = 0
+	}
+	if err := c.incBalance(ic, accFrom, inc); err != nil {
+		return err
+	}
+	if err := ic.dao.PutAccountState(accFrom); err != nil {
+		return err
+	}
+
+	if !isEmpty {
+		accTo, err := ic.dao.GetAccountStateOrNew(to)
+		if err != nil {
+			return err
+		}
+		if err := c.incBalance(ic, accTo, amount); err != nil {
+			return err
+		}
+		if err := ic.dao.PutAccountState(accTo); err != nil {
+			return err
+		}
+	}
+
+	c.emitTransfer(ic, &from, &to, amount)
+	return nil
+}
+
+func (c *nep5TokenNative) balanceOf(ic *interopContext, args []vm.StackItem) vm.StackItem {
+	h := toUint160(args[0])
+	bs, err := ic.dao.GetNEP5Balances(h)
+	if err != nil {
+		panic(err)
+	}
+	balance := bs.Trackers[c.hash].Balance
+	return vm.NewBigIntegerItem(balance)
+}
+
+func (c *nep5TokenNative) mint(ic *interopContext, h util.Uint160, amount int64) {
+	if amount < 0 {
+		panic("negative amount")
+	} else if amount == 0 {
+		return
+	}
+
+	acc, err := ic.dao.GetAccountStateOrNew(h)
+	if err != nil {
+		panic(err)
+	}
+	if err := c.incBalance(ic, acc, amount); err != nil {
+		panic(err)
+	}
+	if err := ic.dao.PutAccountState(acc); err != nil {
+		panic(err)
+	}
+
+	c.totalSupply += amount
+
+	c.emitTransfer(ic, nil, &h, amount)
+}
+
+func (c *nep5TokenNative) burn(ic *interopContext, h util.Uint160, amount int64) {
+	if amount < 0 {
+		panic("negative amount")
+	} else if amount == 0 {
+		return
+	}
+
+	acc, err := ic.dao.GetAccountStateOrNew(h)
+	if err != nil {
+		panic(err)
+	}
+	if err := c.incBalance(ic, acc, -amount); err != nil {
+		panic(err)
+	}
+	if err := ic.dao.PutAccountState(acc); err != nil {
+		panic(err)
+	}
+
+	c.totalSupply -= amount
+
+	c.emitTransfer(ic, &h, nil, amount)
+}
+
+func (c *nep5TokenNative) onPersist(*interopContext) error { return nil }
+
+func (c *nep5TokenNative) toNativeContract(name string) *NativeContract {
```

Comment:
I have refactored this a bit.

This is complicated, because some methods of NEP5 token contract should call methods overriden by NEO contract. This is why `onPersist` is a member.